### PR TITLE
Rewrite profile editor without PromiseRenderer

### PR DESCRIPTION
**Pull request by eatyourgreens**
Rewrites the account profile settings page, replacing the PromiseRenderer component with component state.

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [x] Did you deploy a staging branch? https://edit-profile.pfe-preview.zooniverse.org
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [x] Are the tests passing locally and on Travis?

## Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

File: app/pages/settings/profile.jsx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+  handleMediaChange(type, file) {
+    const { user } = this.props;
+    const errorProp = `${type}Error`;
+    this.setState({ [errorProp]: null });
+    apiClient.post(user._getURL(type), { media: { content_type: file.type }})
+    .then(([resource]) => {
+      putFile(resource.src, file, { 'Content-Type': file.type });
+    })
+    .then(() => {
```
Comment (eatyourgreens) [inline app/pages/settings/profile.jsx]: Why was this block in the original component? What did it do? Can we just delete it?

File: app/pages/settings/profile.jsx [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  handleMediaClear(type) {
+    const errorProp = `${type}Error`;
+    this.setState({ [errorProp]: null });
+    this[`${type}Get`]
+    .then((resource) => {
+      !!resource && resource.delete();
+    })
+    .then(() => {
```
Comment (eatyourgreens) [inline app/pages/settings/profile.jsx]: Ditto for this block. Why was the original code rewriting component methods on-the-fly?

File: app/pages/settings/profile.jsx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  avatarGet() {
+    this.props.user.get('avatar')
+    .then(([avatar]) => {
+      this.setState({ avatar });
+    });
+  }
+
+  headerGet() {
```
Comment (eatyourgreens) [inline app/pages/settings/profile.jsx]: should be `profile_headerGet`

File: app/pages/settings/profile.jsx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      avatar: {},
+      avatarError: null,
+      header: {},
+      headerError: null
+    };
+    this.avatarGet();
+    this.headerGet();
```
Comment (eatyourgreens) [inline app/pages/settings/profile.jsx]: `this.profile_headerGet`

File: app/pages/settings/profile.jsx (hunk 2 above)
Comment (simoneduca) [reply app/pages/settings/profile.jsx]: I have no idea. Do you mean this also doesn't do anything?

File: app/pages/settings/profile.jsx (hunk 1 above)
Comment (srallen) [reply app/pages/settings/profile.jsx]: I originally wrote these files and I'm pretty sure I copied what was being done in the lab for the image media uploads. Maybe this commit is a clue? https://github.com/zooniverse/Panoptes-Front-End/commit/627021199705ff433439812cd69e0a7f5f4667bb

File: app/pages/settings/profile.jsx (hunk 2 above)
Comment (eatyourgreens) [reply app/pages/settings/profile.jsx]: It does do something. It deletes the event handler then forces a re-render, effectively making the button useless after one successful click. But why?

Comment (eatyourgreens) [reply app/pages/settings/profile.jsx]: The original code generated a new event handler on each render.

File: app/pages/settings/profile.jsx (hunk 3 above)
Comment (eatyourgreens) [reply app/pages/settings/profile.jsx]: Or maybe not. I'm confused by my own code.

**Comment by eatyourgreens** [comment]
Rebased, re-staged and, most importantly, re-written. I'm a lot happier with this version.

**Comment by eatyourgreens** [comment]
I'm seeing the avatar being retrieved for both the avatar and profile header. Not sure why that's happening.

**Comment by eatyourgreens** [comment]
Waiting on zooniverse/Panoptes#2528 to fix profile images.

File: app/pages/settings/profile.jsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      return resource;
+    });
+  }
+
+  handleMediaChange(type, file) {
+    const { user } = this.props;
+    const errorProp = `${type}Error`;
+    this.setState({ [errorProp]: null });
+    apiClient.post(user._getURL(type), { media: { content_type: file.type }})
+    .then(([resource]) => {
```
Comment (eatyourgreens) [inline app/pages/settings/profile.jsx]: `resource.src` here is a signed S3 URL, for uploading the file, so we can't store this resource in component state.

File: app/pages/settings/profile.jsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      putFile(resource.src, file, { 'Content-Type': file.type })
+      .then(() => this.getUserResource(type))
```
Comment (eatyourgreens) [inline app/pages/settings/profile.jsx]: Fetch the resource from Panoptes, again, in order to get the public S3 URL.

File: app/pages/settings/profile.jsx [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+import putFile from '../../lib/put-file';
+
+const MAX_AVATAR_SIZE = 65536;
+const MAX_HEADER_SIZE = 256000;
+
+class CustomiseProfile extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      avatar: {},
```
Comment (eatyourgreens) [inline app/pages/settings/profile.jsx]: Uses empty objects for defaults, to avoid `undefined property of null` errors later.

**Comment by eatyourgreens** [comment]
@srallen thanks for your comments. I think that code was there to force an update when the media file changes. Should all be working in the new code now, by fetching the new resource from Panoptes and storing it in component state.

**Comment by eatyourgreens** [comment]
The Panoptes bug is fixed, so I think this is ready for review now @srallen @simoneduca

**Comment by eatyourgreens** [comment]
Added some tests for the initial load of the user images and the two delete buttons.

**Comment by eatyourgreens** [comment]
Rebased to latest master.

File: app/pages/settings/profile.jsx (hunk 7 above)
Comment (simoneduca) [reply app/pages/settings/profile.jsx]: Is that a better approach in general?

File: app/pages/settings/profile.spec.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  });
+
+  it('renders the user background', () => {
+    const avatar = wrapper.find('ImageSelector[src="//zooniverse.org/images/header.jpg"]');
+    assert.equal(avatar.length, 1);
+  });
+
+  it('deletes the avatar when Clear Avatar is pressed', () => {
+    wrapper.find('button').first().simulate('click');
+    setTimeout(() => assert.equal(wrapper.update().state().avatar, {}));
```
Comment (simoneduca) [inline app/pages/settings/profile.spec.js]: @eatyourgreens Is this intended to simulate the avatar resource being an empty object when deleted? If so, shouldn't the test fail if I add`src: 'avatar'` inside `{}`?

Comment (eatyourgreens) [reply app/pages/settings/profile.spec.js]: Yep.

File: app/pages/settings/profile.jsx (hunk 7 above)
Comment (eatyourgreens) [reply app/pages/settings/profile.jsx]: Depends on your code.

File: app/pages/settings/profile.spec.js (hunk 8 above)
Comment (simoneduca) [reply app/pages/settings/profile.spec.js]: It doesn't.

Comment (eatyourgreens) [reply app/pages/settings/profile.spec.js]: Interesting. I'll take another look at this when I'm free. Did you find any clues as to why it isn't working?

Comment (simoneduca) [reply app/pages/settings/profile.spec.js]: I've never seen `setTimeout` used in test so I was checking if it plays well with `wrapper.update()`.

**Pull request merged**